core/src/components/list/list.md.scss

```diff
@@ -23,12 +23,46 @@
   @include border-radius($list-inset-md-border-radius);
 }
 
-.list-md.list-inset ion-item:first-child {
+/**
+ * These selectors ensure the first item in the list
+ * has the correct radius.
+ * We need to consider the following scenarios:
+ 1. The first item in a list as long as it is not the only item.
+ 2. The item in the first item-sliding in a list.
+ * Note that we do not select "ion-item-sliding ion-item:first-of-type"
+ * because that will cause the borders to disappear on
+ * items in an item-sliding when the item is the first
+ * element in the item-sliding container.
+ */
+.list-md.list-inset ion-item:not(:only-of-type):first-of-type,
+.list-md.list-inset ion-item-sliding:first-of-type ion-item {
   --border-radius: #{$list-inset-md-border-radius $list-inset-md-border-radius 0 0};
 }
 
-.list-md.list-inset ion-item:last-child {
-  --border-radius: #{0 0 $list-inset-md-border-radius, $list-inset-md-border-radius};
+/**
+ * These selectors ensure the last item in the list
+ * has the correct radius and border.
+ * We need to consider the following scenarios:
+ 1. The last item in a list as long as it is not the only item.
+ 2. The item in the last item-sliding in a list.
+ * Note that we do not select "ion-item-sliding ion-item:last-of-type"
+ * because that will cause the borders to disappear on
+ * items in an item-sliding when the item is the last
+ * element in the item-sliding container.
+ */
+.list-md.list-inset ion-item:not(:only-of-type):last-of-type,
+.list-md.list-inset ion-item-sliding:last-of-type ion-item {
+  --border-radius: #{0 0 $list-inset-md-border-radius $list-inset-md-border-radius};
+  --border-width: 0;
+  --inner-border-width: 0;
+}
+
+/**
+ * The only item in a list should have a border radius
+ * on all corners.
+ */
+.list-md.list-inset ion-item:only-child {
+  --border-radius: #{$list-inset-md-border-radius};
```

Comment:
This change seems broken to me:

| main | branch |
| ---| ---|
| ![main(Pixel 5)](https://github.com/ionic-team/ionic-framework/assets/6577830/b1d39525-e822-4ec1-baa0-8fdffbebfac0) | ![branch(Pixel 5)](https://github.com/ionic-team/ionic-framework/assets/6577830/c5c15fc9-c284-4929-8a6e-edbce0bcc666) |

Steps to reproduce:

1) Modify `core/src/components/list/test/inset/index.html` to:

```diff
- <ion-content id="content" color="light">
+ <ion-content id="content" color="tertiary">
+   <ion-list inset>
+     <ion-item>
+       <ion-label>Text Items</ion-label>
+     </ion-item>
+   </ion-list>
+
    <ion-list inset>
```

Add the following `style` to the `head`:

```html
<style>
  ion-item {
    --background: red;
  }
</style>
```

Increase the border radius on `md` to see the change better:

```scss
$list-inset-md-border-radius:          10px !default;
```

Reply:
I'm just following what was there before for border radius: https://github.com/ionic-team/ionic-framework/blob/a4551470a9700c3278d0005f083e7a8b6ddd04b5/core/src/components/list/list.md.scss#L26-L33

Demo: https://codepen.io/liamdebeasi/pen/eYxzXjw

Note that the last-child's border radius never showed up because it was formatted incorrectly: 
![image](https://github.com/ionic-team/ionic-framework/assets/2721089/03e6b896-f2c8-4ed3-9bdc-558f2ad1b5bf)

(either everything should have a comma, or nothing should)

Reply:
Hmm okay. I think this might be broken behavior then on `md`. This code used to be required when the items lined up with the top / bottom of the list but since [Material Design added padding](https://m2.material.io/components/lists#specs) to the top and bottom of a list it doesn't line up properly. It looks like we added that in [5 years ago](https://github.com/ionic-team/ionic-framework/commit/e416c23b3b946f34b0e54d4d236991727dc353a2#diff-a62a9570aa3f44e50c23e047c148e994010732cf3af17f9257672d0f68b30436R20) but didn't account for the border radius since item generally has the same background as list. 

[Material Design 3](https://m3.material.io/components/lists/specs#ee01024f-ea3b-47f4-81af-4f90d12691d6) also has this padding at the top and bottom of their list. Should we create a tech debt ticket to take a look at this or add it to our MD3 work? 

It does look like it should still have this code on `ios` though. We also don't seem to have any tests verifying this behavior.